### PR TITLE
[5.3] [WIP] Custom Type Casting

### PR DESCRIPTION
**Pull request by AdenFraser**
This is an early draft of some custom type casting functionality, discussed here: 
https://github.com/laravel/internals/issues/9

The proposed implementation would allow Model Attributes to be cast to custom classes, with optional parameters.

For instance:

``` php
    /**
     * The attributes that should be cast to native types.
     *
     * @var array
     */
    protected $casts = [
        'html_content' => HtmlString::class,
        'money' => '\App\Money,£',
    ];
```

Where the cast class is instantiated with the attribute value as the first parameter, followed by any additional parameters provided.

Assuming a casting class has a __toString() method, this will be used when inserting the attribute back into the database.

A simple custom class example could be automatic casting to HtmlString (useful for WYSIWYG stored content). Creating additional cast classes for Value Objects, such as Money, Currency, Url etc would be as straightforward as creating a class with a __construct and __toString().

``` php
class Money 
{
    protected $value;

    public function __construct($value, $symbol)
    {
        $this->value = $value;
        $this->symbol = $symbol;
    }

    public function withSymbol()
    {
          return $this->symbol.$this->value;
    }

    public function add($value)
    {
          $this->value  = $this->value + $value;

          return $this;
    }

    public function __toString()
    {
         return $this->value;
    }
}
```

This could certainly be made stricter, perhaps with a Castable interface to be implemented on castable objects but it works nicely as above.

**This is an early draft to gather some feedback, thoughts?**

:)

### CODE REVIEW

**Comment by AdenFraser** [comment]
I've noticed this is breaking casting to 'datetime' (a Carbon instance) - fixing now.

**Comment by tomschlick** [comment]
I think there needs to be a way for the cast class to access the model, or at least multiple fields. 

For instance if you had a payments table utilizing the `Money` class example you may have different currencies per row. So it would be nice to have the `amount` field as well as the `currency` field both injected into the class so you can pass back the correct format.

**Comment by AdenFraser** [comment]
I agree that would be useful, I'm not sure how best to implement that without making the process of creating a Casting class more conveluted. Thoughts?

**Comment by tomschlick** [comment]
maybe something like this?

```php

/**
     * The attributes that should be cast to native types.
     *
     * @var array
     */
    protected $casts = [
        'amount' => [Money::class, 'currency'],
        'amount' => '\App\Money,currency',
        'amount' => '\App\Money,£',
    ];
```

the thought is if the parameter == a field name then inject that field, if not then just inject that string.

**Comment by barryvdh** [comment]
See my comment here: https://github.com/laravel/internals/issues/9#issuecomment-214725670

This seems a bit confusing and not in line with other Laravel declarations (eg. routes/validators are `My\Class@method`, or `money:£`

**Comment by barryvdh** [comment]
Also, probably better to use the Laravel App container to build your classes, instead of reflection, if you go this way.

**Comment by AdenFraser** [comment]
@barryvdh Agree with your comments https://github.com/laravel/internals/issues/9#issuecomment-214725670

@tomschlick I think to avoid unexpected conflicts of field values as parameters what you've proposed would need to be more inline with the way Routing and Middleware parameters work - where {field} is injected with it's dynamic value, for instance

```php
protected $casts = [
    'amount' => '\App\Money,{currency}',
];
```

**Comment by AdenFraser** [comment]
I do agree that building the class through the App container is the way to go!

**Comment by tomschlick** [comment]
Yeah 👍  for that.

**Comment by GrahamCampbell** [comment]
> I do agree that building the class through the App container is the way to go!

Note that we cannot do that in an eloquent model. It cannot be coupled like that.

**Comment by JosephSilber** [comment]
What would be the reason of resolving it out of the container? You should only ever cast to value objects, and those shouldn't ever have dependencies.

For example, if you want the ability to convert between different currencies at the current rate, you should have a separate service that handles that; your `Money` class should not be aware of that at all.

**Comment by AdenFraser** [comment]
@JosephSilber I largely agree, although I wonder if there would be a use case where someone would like to resolve an interface implementation of 'Money'. I'm not to sure of the practicality of that though.

File: src/Illuminate/Database/Eloquent/Model.php [hunk 1]
```diff
@@ -2945,6 +2949,76 @@ protected function asTimeStamp($value)
...
+
+    /**
+     * Has a custom cast type been defined for a model attribute and does it exist.
+     *
+     * @param  string $key
+     * @return bool
+     */
+    protected function castExists($key)
+    {
+        if (! $this->castClass($key)) {
```
Comment (lucasmichot) [inline src/Illuminate/Database/Eloquent/Model.php]: return $this->castClass($key);

File: src/Illuminate/Database/Eloquent/Model.php [hunk 2]
```diff
@@ -2945,6 +2949,76 @@ protected function asTimeStamp($value)
...
+        }
+
+        return true;
+    }
+
+    /**
+     * Return the fully qualified custom class.
+     *
+     * @param  string $key
+     * @return string
```
Comment (lucasmichot) [inline src/Illuminate/Database/Eloquent/Model.php]: Return mixed

File: src/Illuminate/Database/Eloquent/TypeCaster/TypeCaster.php [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+    /**
+     * Get a cast type and its parameters for a given attribute.
+     *
+     * @param  string  $attribute
+     * @return array|null
+     */
+    protected function getCast($attribute)
+    {
+
```
Comment (lucasmichot) [inline src/Illuminate/Database/Eloquent/TypeCaster/TypeCaster.php]: // are usually set for empty methods

File: src/Illuminate/Database/Eloquent/Model.php (hunk 2 above)
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Model.php]: mixed should be avoided if possible

Comment (lucasmichot) [inline src/Illuminate/Database/Eloquent/Model.php]: In this case adding |void to return will be fine

**Comment by AdenFraser** [comment]
I'll be attempting to wrap this up later this week, I'm currently transitioning between machines so a few things have had to go on the backburner.

**Comment by halaei** [comment]
If it is OK to use PHP serialization, It will be possible to support inheritance as well. Please take a look at  #13511

**Comment by taylorotwell** [comment]
Where are we at with this? :)

**Comment by AdenFraser** [comment]
Sorry been hectic with a change in jobs. Will sink my teeth into this
properly this week.

Happy to open up write access to my fork (if that's possible?) If others
would like to assist in fleshing it out.

Definitely need to find time for this as I hope it can make it into 5.3 ;)

**Comment by taylorotwell** [comment]
I have an alternative implementation of this that I think addresses all concerns fairly elegantly. (Resolving stuff from IoC, mutating multiple fields into one aggregate value object, etc.) I'll try to get it pushed up as a PR today so y'all can take a look and see what you think.

**Comment by taylorotwell** [comment]
While I have an implementation, I am still curious why people would prefer this approach over just defining `getFooAttribute` and `setFooAttribute` on their model and creating the value objects that way. What advantage would this approach give?

**Comment by JosephSilber** [comment]
Using accessors and mutators means you have to create a set of those for every single attribute that you want it in, in every model.

Using casting means you define the functionality once, then simply add the attribute names to the casts array.

Much less boilerplate this way.

**Comment by AdenFraser** [comment]
@taylorotwell Awesome, I look forward to seeing what you decided to put together!

**Comment by halaei** [comment]
I have a question. If we change the casted object, will the model get affected as well? e.g:
```php
$model->money = '10$';
$model->money->add(10)
echo $model->money;
```
is `$model->money` now 20$ or still 10$?

**Comment by AdenFraser** [comment]
When the save action is used, it would be expected that your casted option
has a way of conversion from object back to string for database storage.

Presumably most people would want the new value to be stored in the
database, the same as money being an integer, adding 10 and saving would do.

**Comment by halaei** [comment]
@AdenFraser sorry don't find my answer. I **wasn't** talking about the saved value in the database, I am considering the in memory values. Let me put it this way, if you read `$model->money` twice, will you get *the same* object for the second time or will it be a new Money object that is equal to the previous one, but not the same. Or simply, will this assertion pass?
```php
    assert($model->money === $model->money);
```

**Comment by AdenFraser** [comment]
Once the value is casted, you will receive an object. So you'd have to
assert that $model->money is equal to a Money Object which has a value
property now set to 20.

In memory it would behave the same as accessory and mutations- the object
is what you would receive when you attempt to access the money property on
your model.

**Comment by taylorotwell** [comment]
Yeah I think that value object casting would need to be cached and only done once per model like Hamid is saying.

**Comment by taylorotwell** [comment]
https://github.com/laravel/framework/pull/13706

Let's take up discussion here.

**Pull request closed without merging**